### ISSUE TITLE
Lease Resource Id Restriction

### ISSUE THREAD

**Issue by andrewdmoreno** [f1x3d/DistributedLeaseManager#5]
Out of curiosity, was there a specific reason that the `ResourceId` was implemented as a `Guid` and not just a `string`? Let's say we have some deterministic string value, a business key or the like, that we would like to lock on we are unable to. 

Example: We have an item id that has an alpha-numeric value as an identifier and we would like to get a lock on that item before updating stock levels for that part.

I know Cosmos DB and probably other implementations have restrictions on what characters can be in the `Id` but still feels like it would be useful to have this capability knowing those restrictions.

Thoughts?

**Comment by f1x3d** [comment]
Hey @andrewdmoreno,

First of all, thanks for spending time seeing how this library works!

You have a good point that currently it is required to provide some GUID `ResourceId` even if there is no value like that on the business side.

As you were able to tell, the current implementation was initially tailored to Cosmos DB, hence the GUID type for an ID was chosen. That being said, the `ResourceId` on its own doesn't have any specific purpose other than to be used in conjunction with `ResourceCategory` to form a unique lease identifier. 

This means that it's possible to implement it in a way to have a unique constraint on the `ResourceCategory + ResourceId` pair while keeping the underlying data storage ID as an auto-generated value. I will consider making these improvements and creating the v2 version of the library since these are likely to be breaking changes.

As a workaround for now, you can make the string value part of your `ResourceCategory` while keeping the `ResourceId` static/empty, e.g.

```csharp
await leaseManager.TryAcquireLease($"StockUpdate-{item.Sku}", Guid.Empty, TimeSpan.FromSeconds(5))
```

That makes it look a bit weird from a semantic standpoint but it should work as expected.

Let me know if that helps or if you encounter any other issues!

**Comment by andrewdmoreno** [comment]
Thank you for the quick response! Yes that makes sense and thank you for the suggestion. I overlooked keeping the `Id` static and varying the partition key. May look a little funny to have the id constant for all the items but will still work.

I am currently investigating the differences between your implementation and the sample implementation proposed by the Cosmos Db team on their [design patterns blog](https://devblogs.microsoft.com/cosmosdb/azure-cosmos-db-design-patterns-part-4-global-distributed-lock/). 

Based on my research so far it seems the major difference so far is that they attempted to include Fence token support. They also maintain the concept of a lease on a lock as a separate construct than the lock itself (although I'm not sure if that's really needed). As it is just a sample though there is no nuget package and the api surface seems like it could use some polishing before using in an application. Yours on the other hand is much more client friendly and ready to use in an application, thanks for putting it together!

**Comment by f1x3d** [comment]
Thanks for linking the Microsoft sample, haven't seen it before.

The model separation may be coming from their [Cosmos Change Feed lease implementation](https://github.com/Azure/azure-cosmos-dotnet-v3/tree/6d603f54bdfee42b2768e7407fdc4eb56fcc8271/Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/LeaseManagement), which to me feels more like a [semaphore](https://en.wikipedia.org/wiki/Semaphore_(programming)) rather than a lock, since there could be multiple feed listeners per container. Perhaps they used a stripped-down version of that as an example of locking. Not sure it makes any difference for the single consumer case.

Glancing at the rest of their code, it appears the two main differences are the owner checks and the fence tokens.

The owners were omitted from my library based on the assumption that you wouldn't try to "steal" someone else's lease by writing code like this:

```csharp
var lease = DistributedLeaseAcquisitionResult.Success(
    new DistributedLeaseRepository(/*...*/),
    new DistributedLease
    {
        ResourceCategory = "ImportantOperation",
        ResourceId = Guid.Parse("b82eac94-888c-477d-9149-1b585abcb6f6"),
    });

await lease.DisposeAsync(); // Potential steal of the lease
```

In other words, if you don't create the `DistributedLeaseAcquisitionResult` object manually, then you are the lease owner as long as you have the corresponding lease object instance.

The fence token is helpful when protecting against delayed operations being executed in cases where an old lease owner tries to use it after its expiration. Since expiration checks in `DistributedLeaseManager` don't rely on the backing storage auto-removal features, like TTL in Cosmos DB, the `ExpirationTime` property can act as a fence token as it is updated whenever a new lease is issued.

**Comment by andrewdmoreno** [comment]
Let me know if you would like any contributions in regards to this. If you give me direction on how you prefer it implemented I can make sure to follow suit.

I looked at the surface area and really it just impacts `DistributedLease` which isn't directly exposed externally to callers as it's not part of the `IDistributedLeaseAcquisitionResult` interface.

**Comment by f1x3d** [comment]
Always open to accepting contributions!

Agreed, there's not going to be a lot of changes. A rough plan:
- Change the `ResourceId` property type from `Guid` to `string`
- Update all references to accept a `string` as well
- Make sure the ID values for different storage types are being populated correctly with the new data type, i.e. an `id` in Cosmos DB, a table primary key in EF Core and a file name in Blob Storage. The EF Core ID can stay as a `Guid` since it's a shadow property anyway
- Make sure the unique constraint on the `ResourceCategory + ResourceId` pair for all storage types is intact with the new data type
- Make sure you can still find leases created with the current code version when using the new code version for all storage types. As a way to test this, you can create a lease with a year duration using the current code version and then try to acquire the same lease after the data type changes, which should fail
